### ISSUE TITLE
Related inputs do not work when one of them is not answered

### ISSUE THREAD

**Issue by C0untC0unt** [maths/moodle-qtype_stack#195]
We already mentioned this problem in another issue but we think it is worth a new one to give a suggestion/demand to deal with it.  #157
When the student is asked to fill in two (or more) inputs within one single question and ans2 depends on ans1, the specific feedback does not appear if one of the input-boxes stays empty.

**Problem:** If the specific feedback is meant to differ depending on the combination of ans1 and ans2, the only chance to solve this is by using one single PRT for both inputs. BUT, as long as one of the inputs isn't answered this PRT does not trigger.
Using 3 PRTs for two inputs does not help with this problem because the feedbacks to the single answers can not be related. That would only give a "Right" or "Wrong" to each answer but not an sophisticated response to the combination of them.

**A solution** to this porblem would be, to make it possible to check within a PRT, if the input-boxes are empty or not. That's way more easier and efficient than the 3-PRT approach.

**Comment by sangwinc** [comment]
This behavior is not a bug, but it was a conscious design decision.

The internals assign a "status" to each input, including NULL, VALID, INVALID, SCORE etc.  Only when all inputs are at status "SCORE" will the PRT be executed.  We made this choice because of the way in which penalties are applied, and how all this interacts with page navigation.  

It isn't as simple as just having PRTs (possibly ?) accept empty inputs.   

Maxima will not accept an empty assignment, so we would have to choose something like "null".  Then we would have to distinguish between a student typing "null" and a student leaving something blank, etc.   Typical problem with this situation!

I've always designed STACK from the academic perspective, and tried to limit the technical constraints.  Given there is a solution to this problem (3 PRTs) I'm interested to know (1) what questions do you need to author with STACK and why to you really need this, (2) how often do you do this?

As a volunteer maintainer, I am very happy to change/extend functionality with additional options, but there is a cost in terms of my time.  I need to be satisfied I'm not about to spend my weekend making it more efficient for you (just you?) to write one question!   :-)

Chris

**Comment by aharjula** [comment]
No. The PRT system already checks whether the boxes are filled and filled validly before acting at all and this logic has many good reasons to be like it is. Some comments:

- You cannot give feedback about a set of inputs if they are not all valid, which is logical. 
- You can still attach a PRT to each of those inputs and give feedback about that single input or subset of inputs, but naturally that PRT cannot know about inputs it has not been attached to.
- The PRT system ties tightly to the penalty systems present in many question behaviours and if the PRTs were to activate when there are empty inputs we would probably generate penalties for incomplete attempts.
- The PRT system as is has been such since the time STACK got multiple input fields and multiple PRTs and changing this fundamental behaviour of it would cause serious problems. Probably even forks.

What you basically want is to define a way of accessing input fields inside the PRT without making those input fields a requirement for that PRT to be executed, which is actually doable in my branch of STACK given certain assumptions about the order of those PRTs. But the reason you want is wrong and you should seriously think about this a bit more. In my opinion you should not add conditions about input fields not affecting directly to the PRT inside the PRT, these cause trouble and will eventually make your questions somewhat convoluted and hard to maintain. The correct way is still, has always been and will be in the future even with my version of STACK as follows:

- PRT-feedback about a combination of inputs activates only when all those inputs have valid values.
- If needed you can add additional PRTs to cover subsets of those inputs. Noting that those feedbacks will fire even when all those inputs are filled.
- You may choose yourself how you give points from those PRTs if the answer requires that specific sets of inputs have been filled you obviously give points only from the PRTs that tie to those sets.

Chris types faster... I am completely on his side on this, the question model using this way of PRT activation is simple yet extremely expressive and adding the complexity you require does not provide enough benefits (if any). And I would know about complexity, considering the mad things I have done to STACK (Chris maybe you want to go and look at my new sample questions, recursive and infinite and so on).

**Comment by C0untC0unt** [comment]
Wow, you guys are fast. :-)

Ok, I understand your concerns. Unfortunately, I am running out of time to go into detail right now. But I will do so in the next days and beforehand I will have a look on Matti's branch of STACK.
**"What you basically want is to define a way of accessing input fields inside the PRT without making those input fields a requirement for that PRT to be executed.."** That seems to be a description to what I ask for.
Thanks to both of you so far.

**Comment by aharjula** [comment]
Just a warning, if you do this with my branch you will need to learn to do things with state variables and people here think that all "state is evil". You will also need to solve by yourself how you would do the thing you want with those variables, I assure you that it is possible but will require at least one extra PRT for each input and you will need to figure out how to handle the cases where inputs return to empty values. I won't help with that as I think this is a bad idea. The branch has some docs about state and some sample questions that should show give some ideas about how they work.
